File: hcs_toolbox_py/lpic2hpic.py
"""
Upgrade Linear Proportional-Integral Control (LPIC) to Homogeneous
Proportional-Integral Control (HPIC).

Original MATLAB: lpic2hpic.m

The upgrade:
    Linear:   u = K*x + v,   dv/dt = Ki*x
    HPIC:     u = uh + v,    dv/dt = ui

    uh = K0*x + nx^(1+mu) * (K - K0) * px
    ui = nx^(1+2*mu) * Ki_new * px / (px' * P * Gd * px)

where px = d(-ln(nx))*x is the homogeneous projection.
"""

import numpy as np
from scipy.linalg import sqrtm
from .block_con import block_con


def lpic2hpic(A, B, K, Ki):
    """
    Upgrade a linear PI controller to a Homogeneous PI Controller (HPIC).

    Parameters
    ----------
    A : (n, n) ndarray
        System matrix.
    B : (n, m) ndarray
        Control matrix (must have full column rank).
    K : (m, n) ndarray
        Proportional gain matrix. A + B*K must be Hurwitz.
    Ki : (m, n) ndarray
        Integral gain matrix.

    Returns
    -------
    K0 : (m, n) ndarray
        Feedback gain of the linear component of HPC.
    G0 : (n, n) ndarray
        Matrix defining the generator of dilation: Gd = I + mu*G0.
    P : (n, n) ndarray
        Positive definite matrix defining the canonical homogeneous norm.
    Ki_new : (m, n) ndarray
        Redesigned integral gain for HPIC.
    mu_min : float
        Minimal admissible homogeneity degree.
    mu_max : float
        Maximal admissible homogeneity degree.
    """
    tol = 1e-5
    n = A.shape[0]
    m = B.shape[1]

    # --- Controllability check ---
    U_ctrb = B.copy()
    tA = A.copy()
    k_ctrb = 0
    if np.linalg.matrix_rank(U_ctrb) == n:
        k_ctrb = 1
    i = 1
    while (k_ctrb == 0) and (i <= n - 1):
        i += 1
        U_ctrb = np.hstack([U_ctrb, tA @ B])
        if np.linalg.matrix_rank(U_ctrb) == n:
            k_ctrb = i
        tA = tA @ A

    if k_ctrb == 0:
        raise ValueError("lpic2hpic: The system is not controllable.")

    if np.linalg.matrix_rank(B) < m:
        raise ValueError("lpic2hpic: The control matrix must have full column rank.")

    # --- Stability margin check (augmented system) ---
    # Check eigenvalues of [A+B*K  B; Ki  0]
    M_aug = np.block([[A + B @ K, B],
                      [Ki, np.zeros((m, m))]])
    rho = -max(np.real(np.linalg.eigvals(M_aug)))
    if rho < tol:
        raise ValueError("lpic2hpic: The linear control system does not have "
                         "sufficient stability margin.")
    rho = min(rho, 1.0)

    # --- Special case: B is square and invertible ---
    if np.linalg.matrix_rank(B) == n:
        K0 = -B.T @ np.linalg.inv(B @ B.T) @ A
        G0 = np.zeros((n, n))
        P = np.eye(n)
        Ki_new = Ki.copy()
        mu_min = -0.5
        mu_max = 1.0
        return K0, G0, P, Ki_new, mu_min, mu_max

    # --- Block decomposition ---
    T, nt = block_con(A, B)
    if isinstance(T, int) and T == 0:
        raise ValueError("lpic2hpic: Block decomposition failed.")

    Anew = T @ A @ np.linalg.inv(T)
    k = len(nt)

    # Build block row start indices (0-based)
    n_ind = [0]
    s = 0
    for i in range(k - 1):
        s += nt[i]
        n_ind.append(s)

    Bnew = T @ B
    B0 = Bnew[n_ind[-1]:n, :m]
    K0 = -B0.T @ np.linalg.inv(B0 @ B0.T) @ Anew[n_ind[-1]:n, :n]
    K0 = K0 @ T

    # --- Compute G0 ---
    vG0 = []
    for i in range(k):
        vG0.extend([(k - 1 - i)] * nt[i])
    G0 = -np.linalg.inv(T) @ np.diag(vG0) @ T

    # --- Compute P via Lyapunov-like equation ---
    At = A + B @ K
    I_n = np.eye(n)
    I_m = np.eye(m)

    P12 = -0.5 * B @ np.linalg.inv(B.T @ B)

    # Check eigenvalues (uses A, NOT At = A+B*K — matches MATLAB lpic2hpic line 123)
    lambdas_check = np.real(np.linalg.eigvals(
        -P12 @ Ki - Ki.T @ P12.T + A.T @ P12 @ P12.T @ A))
    lam = max(0.0, -np.min(lambdas_check))

    # Solve Lyapunov equation for P1 (reshaped)
    W1 = np.kron(I_n, At.T) + np.kron(At.T, I_n)
    Q1 = (2.0 + lam) * I_n + A.T @ P12 @ P12.T @ A - P12 @ Ki - Ki.T @ P12.T
    zet1 = -Q1.flatten('F')
    v_P1 = np.linalg.solve(W1, zet1)
    P = v_P1.reshape((n, n), order='F')

    if max(np.real(np.linalg.eigvals(P))) < tol:
        raise ValueError("lpic2hpic: Upgrade impossible (P not positive definite).")

    if max(np.real(np.linalg.eigvals(P @ At + At.T @ P))) > -tol:
        raise ValueError("lpic2hpic: Upgrade impossible (Lyapunov condition fails).")

    # --- Compute Ki_new ---
    # Solve for P2 from coupled equations:  Ki'*P2 = -B'*P  and  P2*Ki = -P*B
    # Using least squares on the stacked system
    W3 = np.kron(Ki.T, I_m)    # kron(Ki', I_m)
    zet3 = (-B.T @ P).flatten('F')
    W4 = np.kron(I_m, Ki.T)    # kron(I_m, Ki')
    zet4 = (-P @ B).flatten('F')

    W_stacked = np.vstack([W3, W4])
    zet_stacked = np.hstack([zet3, zet4])
    v_P2, _, _, _ = np.linalg.lstsq(W_stacked, zet_stacked, rcond=None)
    P2 = v_P2.reshape((m, m), order='F')
    P2 = (P2 + P2.T) / 2.0

    if max(np.real(np.linalg.eigvals(P2))) < tol:
        raise ValueError("lpic2hpic: Upgrade impossible (P2 not positive definite).")

    Ki_new = -np.linalg.inv(P2) @ B.T @ P

    # --- Compute admissible mu range ---
    sqrtP = sqrtm(P)
    inv_sqrtP = np.linalg.inv(sqrtP)
    M_sym = sqrtP @ G0 @ inv_sqrtP + inv_sqrtP @ G0.T @ sqrtP
    lambdas = np.real(np.linalg.eigvals(M_sym))
    lambda_min = np.min(lambdas)
    lambda_max = np.max(lambdas)

    if lambda_max > tol:
        mu_min = max(-0.5, -1.0 / lambda_max + tol)
    else:
        mu_min = -0.5

    if lambda_min < -tol:
        mu_max = min(1.0 / k, -1.0 / lambda_min)
    else:
        mu_max = 1.0 / k

    return K0, G0, P, Ki_new, mu_min, mu_max

File: HCS_Toolbox_ver02/lpic2hpic.m
function [K0, G0, P, Ki_new, mu_min, mu_max]=lpic2hpic(A,B,K,Ki)
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% The function [K0,G0,P,Ki_new,mu_min,mu_max]=lpc2hpc(A,B,K,Ki) upgrades
%% the linear proportional-integral controller (LPIC)
%% 
%% u=K*x+v,  dv/dt=Ki*x      (x - system state)   
%%
%% to the Homogeneous Proportional-Integral Controller (HPIC): 
%%                                                             
%%        u = uh + v,  dv/dt=ui   
%%
%%  uh=K0*x + nx^(1+mu)*(K-K0)*px            (nx - homogeneous norm of x)
%%  
%%  ui=nx^(1+2*mu)*Ki_new*p(x)/(px'*P*Gd*px)      
%%
%%  where  px=d(-log(nx))*x - homogenenous projection                                                                    
%% 
%%
%% The input parameters:                                             
%%      A  - system matrix (n x n)                                       
%%      B  - control matrix (n x m)       
%%      K  - proportional gain matrix (m x n) of LPIC s.t. A+B*K - Hurwitz  
%%      Ki - integral gain matrix (m x n) of LPIC 
%%
%% The output parameters:                                             
%%      K0 - feedback gain (m x n) of a linear part of HPC controller                                          
%%      G0 - matrix (n x n) that  defines the generator Gd=eye(n)+mu*G0  
%%           of the dilation d(s)=expm(s*Gd)                   
%%       P - positive definite matrix which defines the canonical        
%%           homogeneous norm nx induced by the norm sqrt(x'*P*x)
%%  Ki_new - redesigned gain of integral component of HPIC
%%  mu_min - minimal admissible degree of HPIC  (mu_min<=mu)  
%%  mu_max - maximal admissible degree of HPIC  (mu<=mu_max)
%%
%%
%% WARNING: the upgrade neends the full column rank of the control matrix B 
%%
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
K0=0; G0=0; P=0; Ki_new=0; mu_min=0; mu_max=0;
tol=0.00001; %tolerance

%1) Check the parameters

sA=size(A); sB=size(B); sK=size(K);

if sA(1)~=sA(2) disp('Error: the system matrix must be square'); return;
end;
if sB(1)~=sA(2) disp('Error:  dimensions of system  and control matrices must agree');return;
end;
if sK(1)~=sB(2) disp('Error:  dimensions of control and gain matrices must agree');return;
end;
n=sB(1); m=sB(2);

%controllability check
U=[B];tA=A;
k=0;
if rank(U)==n k=1;
end;i=1;
while (k==0) && (i<=n-1)
    i=i+1;
    U=[U tA*B];
    if rank(U)==n k=i; end;
    tA=A*tA;
end;
if k==0 disp('Error: The system is not controllable'); return;
end;
       
if rank(B)<m disp('Error: The control matrix must have full column rank.'); return; end;

rho=-max(real(eig([A+B*K B; Ki zeros(m,m)])));

if rho<tol disp('Error: The linear control system does not have a sufficient stability margin.'); return; end;
rho=min(rho,1);

if rank(B)==n 
   K0=-B'*inv(B*B')*A;
   Gd0=zeros(n,n);P=eyey(n);
   Ki_new=Ki;
   mu_min=-0.5; mu_max=1;
   return;
end;   
    

%%%%%%%%%%%%%%%%%%%%%
% Design of Gd using block decomposition
%%%%%%%%%%%%%%%%%%

[T nt]=block_con(A,B);
if T==0 K0=0; K=0;P=0; Gd=0; 
    disp('Error: A block decomposition is impossible.'); 
      return; end;
  
Anew=T*A*inv(T);
k=size(nt,1); 
%creation of an array of indecies
s=1;
n_ind=[s];
for i=1:k-1
s=s+nt(i);
n_ind=[n_ind; s];
end;

Bnew=T*B;
B0=Bnew(n_ind(k):n,1:m);
K0=-B0'*inv(B0*B0')*Anew(n_ind(k):n,1:n);
A0=Anew+Bnew*K0;
K0=K0*T;

vG0=[];
for i=1:k
 vG0=[vG0 (k-i)*ones(1,nt(i))];   
end;
G0=-inv(T)*diag(vG0)*T;

Z1=inv(B'*B)*B';
Z0=B*Z1;
At=A+B*K;


I_n=eye(n);I_m=eye(m);
P12=-0.5*B*inv(B'*B);

lambdas=real(eig(-P12*Ki-Ki'*P12'+A'*P12*P12'*A));
lam=max(0,-min(lambdas));
W1=[kron(I_n, At')+kron(At',I_n)];
Q1=(2+lam)*I_n+A'*P12*P12'*A-P12*Ki-Ki'*P12'; 
zet1=-Q1(:);
v_P1=linsolve(W1,zet1);
P=reshape(v_P1,n,n);  

if max(real(eig(P)))<tol disp('Error: The upgrade is impossible.'); return; end;

if max(real(eig(P*At+At'*P)))>-tol disp('Error: The upgrade is impossible.'); return; end;

W3=[kron(Ki',I_m)];
zet3=-B'*P;zet3=zet3(:);
W4=[kron(I_m,Ki')];
zet4=-P*B;zet4=zet4(:);
v_P2=linsolve([W3;W4],[zet3;zet4]);
P2=reshape(v_P2,m,m); 
 
P2=(P2+P2')/2;

if (max(real(eig(P2)))<tol)     disp('Error: The upgrade is impossible.'); return; end;

Ki_new=-inv(P2)*B'*P;

lambdas=real(eig(sqrtm(P)*G0*inv(sqrtm(P))+inv(sqrtm(P))*G0'*sqrtm(P)));
lambda_min=min(lambdas);
lambda_max=max(lambdas);
if lambda_max>tol mu_min=max(-1/2,-1/lambda_max+tol);
else mu_min=-1/2;
end;
if lambda_min<-tol mu_max=min(1/k, -1/lambda_min);
else mu_max=1/k;
end;



